### DISCUSSION TITLE
Overwriting/Updating a module in ShotGrid

### DISCUSSION THREAD

**Discussion by fabiangeisler** [cpenv/cpenv#55] [Q&A] [answered]
If I update/overwrite a module in ShotGrid, will that be (re-)downloaded by an artist that already has this module in his local cache?
Simple example: I created module `test-v1.0` and upload it to SG. Artist xyz gets it in his local module cache, because it was specified in one of his software environment. I add an environment variable that I forgot to the module and publish it again as `test-v1.0` to SG. Will that refreshed module also be updated in the artist cache?

**Comment by danbradham** [comment, marked as answer]
This is another feature I'd like to support. Currently I release a new module any time there needs to be an update or patch.

In an ideal world we would store a timestamp or hash of the module locally to compare against the module in SG. If the modules didn't match we could redownload.

**Reply by fabiangeisler** [reply]
Thanks for the info! Definitely a +1 from me for having that feature.

**Comment by fabiangeisler** [comment]
Was thinking about this some more and really think that this is something essential to put into `cpenv`. Re-uploading fixed modules with different version numbers feels like a bad workaround to me.
- `hash` - would be great for detecting any change at all, about the module
- `timestamp` - would be great for seeing the last publish date and manipulating it by hand if need be.
- For a ShotGrid repo we could also check if the attachment ID of the uploaded files on a module has changed. Sounds fairly easy to implement but is really SG specific. The other solutions sound a bit better since they could work for other types of repos as well. 
I am leaning towards `timestamp`, because I like playing dirty tricks with my tools 😄.

**Reply by danbradham** [reply]
Yeah, all those approaches could work. I think I'll add a couple of methods to the Repo classes. One method would provide an identifier (hash, timestamp, whatever) for a module when it is localized and another method would validate the identifier prior to activation. This way different Repos may provide different validation methods.

I created an issue to track this and hope to tackle it within the next few of weeks.
#63
